src/order/grade.lean

```diff
@@ -0,0 +1,334 @@
+/-
+Copyright (c) 2022 Yaël Dillies, Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Violeta Hernández Palacios, Grayson Burton, Vladimir Ivanov
+-/
+import data.nat.interval
+import data.nat.succ_pred
+import order.atoms
+
+/-!
+# Graded orders
+
+This file defines graded orders, also known as ranked orders.
+
+A graded order is an order with a bottom element in which every element has some finite "height",
+that corresponds to how many elements you need to get down by to reach `⊥`.
+
+## Main declarations
+
+* `grade_order`: Graded orders.
+* `grade`: The grade of an element.
+* `grade_max_order`: Graded orders with maximal elements. All maximal elements have the same grade.
+* `max_grade`: The maximum grade in a `grade_max_order`.
+* `order_embedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## Implementation notes
+
+The standard approach is to define graded orders as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99).
+
+However, this means that graded orders all have a top and that we must reconstruct the grade of an
+element by how many elements we need to go down to reach `⊥`. Unpractical, really.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+open finset nat order_dual
+
+variables {α β : Type*}
+
+/-- A graded order is an order equipped with a grade function which tells how far a given element is
+away from the minimal elements. Precisely, `grade a` is the height of `a` in the Hasse diagram of
+`α`. -/
+class grade_order (α : Type*) [preorder α] :=
+(grade : α → ℕ)
+(grade_strict_mono : strict_mono grade)
+(grade_of_is_min ⦃a : α⦄ : is_min a → grade a = 0)
+(grade_of_covers ⦃a b : α⦄ : a ⋖ b → grade a + 1 = grade b)
+
+/-- A graded top order is a graded order with a maximal grade (NOT a maximal *element*). -/
+class grade_max_order (α : Type*) [preorder α] extends grade_order α :=
```

Comment:
Is this class referred to anywhere in the literature? Or is it only here for your `order_dual` instance?

Reply:
Can you give a reference for this?

Reply:
I refer you to Engel again. That's his definition of "graded poset". `grade_order` is in contrast something intermediate between hi "ranked poset" and "graded poset".

Reply:
Engel's definition additionally insists that the posets are finite, however, which you do not, so this doesn't match the definition of "graded poset" either.

Reply:
Ah but that's a completely different thing. It's a matter of mixin vs extension. `fintype` acts very well as a finiteness mixin so there was no need to extend it.

Reply:
Sure, but this means the definition still doesn't match: we don't have a reference that this is the sensible definition for infinite posets.

Reply:
This instance is precisely the one we'd need for abstract polytopes, or more generally face / exposed face lattices of convex sets, if that's of any value.